packages/react-reconciler/src/ReactFiberWorkLoop.js

```diff
@@ -866,8 +873,14 @@ function finishConcurrentRender(
         }
 
         const nextTime = getNextRootExpirationTimeToWorkOn(root);
-        if (nextTime !== NoWork && nextTime !== expirationTime) {
-          // There's additional work on this root.
+        if (
+          nextTime !== NoWork &&
+          nextTime !== expirationTime &&
+          (!enableTrainModelFix || nextTime > Idle)
```

Comment:
Could put this check into `getNextRootExpirationTimeToWorkOn` instead? Then it will work with `pingSuspendedRoot`, too.

Reply:
More specifically, `getNextRootExpirationTimeToWorkOn` should never return `Never` if the root is suspended, which we can check by looking at `root.firstSuspendedTime` and `root.lastSuspendedTime`.

Reply:
Are firstSuspendedTime and lastSuspendedTime reliable for this type of check? I think they just say that at some point this time was suspended, but it doesn't say that there is still work left on that level (e.g. if the updated component was deleted).

Also `pingSuspendedRoot` doesn't call `getNextRootExpirationTimeToWorkOn` so it wouldn't fix the issue that the finished work gets blown away.

Reply:
We clear `firstSuspendedTime` and `lastSuspendedTime` whenever we commit, so I think it is reliable for this check. Those fields are also how `pingSuspendedRoot` bails out of a ping early, if the pinged time has already finished.

> Also `pingSuspendedRoot` doesn't call `getNextRootExpirationTimeToWorkOn`

It calls `ensureRootIsScheduled` which in turn calls `getNextRootExpirationTimeToWorkOn`. All render passes begin by calling `getNextRootExpirationTimeToWorkOn` so  if you fix it there then you fix it for all cases.

Reply:
To be clear I think we'd still need to remove those lines in `pingSuspendedRoot`. Might be able to delete them? I think they are outdated.

Reply:
I don't think it's that simple. Currently, `Never` can suspend. We also set `lastPingedTime` to `Never` to unblock such work. When we do that, we've lost whether that root was also pinged at a higher priority than that. So if `lastPingedTime` is `Never` then we can't safely return `NoWork` because it could've pinged at higher priority.

Reply:
Actually I can’t seem to write a failing test for this.

If we start another render at higher pri then that unsuspends the Never pri so when the ping comes in it’s no longer suspended and so lastPingedTime doesn’t update.

Maybe this isn’t an issue after all.

That said, Never probably should never suspend the commit.

Reply:
Actually I can’t seem to write a failing test for this.

If we start another render at higher pri then that unsuspends the Never pri so when the ping comes in it’s no longer suspended and so lastPingedTime doesn’t update.

Maybe this isn’t an issue after all.

That said, Never probably should never suspend the commit.